numpy/core/src/multiarray/mapping.c

```diff
@@ -1783,35 +1780,35 @@ array_assign_subscript(PyArrayObject *self, PyObject *ind, PyObject *op)
         return -1;
     }
 
-    /* Single field access */
-    if (PyDataType_HASFIELDS(PyArray_DESCR(self))) {
-        if (PyString_Check(ind) || PyUnicode_Check(ind)) {
-            PyObject *obj;
+    /* field access */
+    if (PyDataType_HASFIELDS(PyArray_DESCR(self)) &&
+            obj_is_string_or_stringlist(ind)) {
+        PyObject *obj;
+        static PyObject *indexfunc = NULL;
 
-            obj = PyDict_GetItem(PyArray_DESCR(self)->fields, ind);
-            if (obj != NULL) {
-                PyArray_Descr *descr;
-                int offset;
-                PyObject *title;
+        if (!(PyString_Check(ind) || PyUnicode_Check(ind))) {
```

Comment:
Is this Py3k save? It might be, we may have some of these macro'd (in which case some of the above can be simplified).
Could you possible add a test that tests for these type of unicode vs. string stuff (assuming they do not already exist)?